language/move-prover/bytecode-to-boogie/src/driver.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::cli::{abort_on_error, Options, INLINE_PRELUDE};
+use crate::translator::{BoogieTranslator, FunctionInfo, ModuleInfo};
+use bytecode_source_map::source_map::SourceMap;
+use bytecode_verifier::VerifiedModule;
+use ir_to_bytecode::compiler::compile_module;
+use ir_to_bytecode::parser::ast::{Loc, ModuleDefinition};
+use ir_to_bytecode::parser::parse_module;
+use itertools::Itertools;
+use libra_types::account_address::AccountAddress;
+use log::{debug, error, info};
+use std::fs;
+use std::process::Command;
+
+/// Content of the default prelude.
+const DEFAULT_PRELUDE: &[u8] = include_bytes!("bytecode_instrs.bpl");
+
+/// Represents the driver for translation.
+pub struct Driver<'app> {
+    /// Options passed via the cli.
+    options: &'app Options,
+    /// List of module specs.
+    module_infos: Vec<ModuleInfo>,
+    /// List of verified modules.
+    verified_modules: Vec<VerifiedModule>,
+    /// Source position map.
+    source_map: SourceMap<Loc>,
+    /// Generated output.
+    pub output: String,
+}
+
+impl<'app> Driver<'app> {
+    pub fn new(options: &'app Options) -> Self {
+        Driver {
+            options,
+            module_infos: vec![],
+            verified_modules: vec![],
+            source_map: vec![],
+            output: String::new(),
+        }
+    }
+
+    /// Runs standard translation. When this finishes, the generated code is in `self.output`.
+    pub fn run(&mut self) {
+        self.load_modules(&self.options.mvir_sources);
+        self.add_prelude();
+        self.translate_modules();
+        // write resulting code
+        info!("writing boogie to {}", self.options.output_path);
+        abort_on_error(
+            fs::write(&self.options.output_path, &self.output),
+            "cannot write boogie file",
+        );
+        if !self.options.generate_only {
+            self.call_boogie(&self.options.output_path);
+        }
+    }
+
+    /// Runs translation in a test context. Instead of writing output to a file, returns
+    /// a pair of the prelude and the actual translated code.
+    pub fn run_for_test(&mut self) -> (String, String) {
+        self.load_modules(&self.options.mvir_sources);
+        self.add_prelude();
+        // Extract the prelude from the generated output, as we want to return it in a separate
+        // string.
+        let prelude = std::mem::replace(&mut self.output, String::new());
+        self.translate_modules();
+        (prelude, std::mem::replace(&mut self.output, String::new()))
+    }
+
+    /// Compiles the list of mvir modules and adds them to the state. On errors, this
+    /// reports and aborts. Nothing will be added to the output yet.
+    pub fn load_modules(&mut self, file_names: &[String]) {
+        let address = AccountAddress::default();
```

Comment:
TODO here for dealing with modules published at addresses other than the default?